### ISSUE TITLE
Text Classification training accuracy problem in distributed training due to not shuffling samples

### ISSUE THREAD

**Issue by philtrade** [fastai/fastai#2501]
**Describe the bug**
When running imdb text classification (second part of lesson3-imdb)  in distributed mode, accuracy dropped to 0.5 on `fit_one_cycle()`, not the 0.92xx in non-distributed mode.

**Provide your installation details**
```
fastai v1, python 3.7, PyTorch v1.2 and v1.4, 3 cuda GPUs (gtx 1080 ti)
platform      : Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic
```

**To Reproduce**
1. Make sure the imdb classification data, and the fine_tuned encoder are already produced.  This can be done by running `lesson3-imdb` notebook up to the "Classifier" section.
```
~/.fastai/data/imdb/data_clas.pkl
~/.fastai/data/imdb/models/fine_tuned_enc.pth
```

2. Run the following gist:
https://gist.github.com/philtrade/33d3982a8cc0f827b1b4e3d1304a08b9:
`python -m torch.distributed.launch --nproc_per_node=3 ./imdb_classifier_ddp.py`
(note: change `--nproc_per_node=3` to the number of GPUs to use)

Output indicates curiously low accuracy and high validation loss:
```
epoch     train_loss  valid_loss  accuracy  time    
0         0.128430    2.518312    0.499960  03:10     
```

**Expected behavior**
When running the `lesson3-imdb` notebook's classifier section, or the above gist in non-distributed single GPU mode:
`./imdb_classifier_ddp.py`, it should achieve accuracy around 0.92-0.93:

```
$ ./imdb_classifier_ddp.py
local_rank: None
epoch     train_loss  valid_loss  accuracy  time
0         0.260723    0.185862    0.928560  02:19                                                                    
```

**Context, and Triage**
The problem seems to be with `fastai.distributed.OurDistributedSampler` is initialized with `shuffle=False` inherited from the `data.train_dl.init_kwargs['shuffle']`.  For Language Modeling training where `data` is a `TextLMDataBunch` object, `False` is quite suitable, see PR #1934.

But  for text classification learner, `data` is  `TextClasDataBunch` object and it randomizes and shuffles data in `SortishSampler.__iter__()`, despite its `train_dl.init_kwargs['shuffle']` is also `False`.  This randomness/shuffling gives great accuracy improvement in training, as the randomizing `SortishSampler` takes over.  So far so good, until `DistributedTrainer` comes into the picture.
 
In `DistributedTrainer.on_train_being()`, the `SortishSampler` is replaced by `fastai.distributed.OurDistributedSampler`, which was initialized with the `data.train_dl.init_kwargs['shuffle'] == False` value:
https://github.com/fastai/fastai/blob/6f3fec2efaa22c67ad2aabba2ecd09814e448842/fastai/distributed.py#L25-L34

This creates a `OurDistributedSampler` that **does not shuffle in `__iter__()`** and yields poor accuracy (consistent over many epochs).

**Proposed Solution**
Instead of [line 34](https://github.com/fastai/fastai/blob/6f3fec2efaa22c67ad2aabba2ecd09814e448842/fastai/distributed.py#L34) above, create OurDistributedSampler with `shuffle=True`, if the data is a text classification data bunch.

```
        shuffle_train = True if isinstance(self.data,TextClasDataBunch) else shuffle
        self.old_train_dl,self.data.train_dl,self.train_sampler = self._change_dl(self.data.train_dl, shuffle_train)
```
With this, distributed training of imdb classifier achieves accuracy of 0.92xx:
```
python -m torch.distributed.launch --nproc_per_node=3 ./imdb_classifier_ddp.py
...
epoch     train_loss  valid_loss  accuracy  time    
0         0.282722    0.188272    0.930526  03:09     
```
 (*Caveat: 2 GPUs trains 2X slower, and 3 GPUs regain some speedup but still slower than 1 GPU.  I feel that's a different performance issue to be investigated separately.*)

A little monkey-ish, but gives the desirable accuracy.  I'll be happy to whip up a PR for this, if it's an acceptable solution for now. Thanks.

**Comment by sgugger** [comment]
The dynamic aspect of text classification (batches with various lengths) will make it hard to get real performance in a distributed fashion, and what is needed is probably a custom sampler. This is not something we are keen on tackling ourselves as our focus is on v2 right now. A PR with the dirty fix can certainly be merged as a first step.

**Comment by philtrade** [comment]
Agree that v2 is **the** focus.  Will keep the PR minimal.  Thx for pointing out the challenge with batches of different lengths.

**Comment by sgugger** [comment]
Fixed by #2504
